src/editor/history.js

```diff
@@ -0,0 +1,101 @@
+/*
+Copyright 2019 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+export default class HistoryManager {
+    constructor() {
+        this._stack = [];
+        this._newlyTypedCharCount = 0;
+        this._currentIndex = -1;
+        this._changedSinceLastPush = false;
+        this._lastCaret = null;
+    }
+
+    _shouldPush(inputType, diff) {
+        if (inputType === "insertText") {
+            if (diff.removed) {
+                // always append when removing text
+                return true;
+            }
+            if (diff.added) {
+                // only append after every 5th keystroke while typing
+                this._newlyTypedCharCount += diff.added.length;
+                return this._newlyTypedCharCount > 5;
```

Comment:
I'd recommend extracting this magic number out to a constant at the top of the module with a comment explaining it's purpose there.